File: test/integration/ForwardCompatMain.rs
//===----------------------------------------------------------------------===//
//
// Part of the OpenCyphal project, under the MIT licence
// SPDX-License-Identifier: MIT
//
//===----------------------------------------------------------------------===//

//! Runtime forward/backward compatibility test for the generated Rust serializers. See
//! ForwardCompatDriver.c for the full rationale: round-trip tests only ever decode a buffer with the
//! same type version that wrote it, so the delimited version-skew skip is never exercised. This decodes
//! across two versions of a delimited composite (`wire.nar.Inner` has one field; `wire.wid.Inner`
//! appends a second).

use dsdl_generated::wire::nar::holder_1_0::wire_nar_Holder_1_0 as NarHolder;
use dsdl_generated::wire::nar::inner_1_0::wire_nar_Inner_1_0 as NarInner;
use dsdl_generated::wire::wid::holder_1_0::wire_wid_Holder_1_0 as WidHolder;
use dsdl_generated::wire::wid::inner_1_0::wire_wid_Inner_1_0 as WidInner;

fn main() {
    // Forward compatibility: wide writer -> narrow reader skips the appended `y`.
    let wide_in = WidHolder { inner: WidInner { x: 1, y: 2 }, tail: 99 };
    let mut wire = [0u8; 64];
    let n = wide_in.serialize(&mut wire).expect("serialize wide");
    let mut narrow_out = NarHolder::default();
    narrow_out
        .deserialize(&wire[..n])
        .expect("deserialize narrow");
    assert_eq!(narrow_out.inner.x, 1, "forward-compat inner.x");
    assert_eq!(narrow_out.tail, 99, "forward-compat tail");
    println!("fwdcompat_ok");

    // Zero extension: narrow writer -> wide reader zero-fills the absent `y`.
    let narrow_in = NarHolder {
        inner: NarInner { x: 5 },
        tail: 77,
    };
    let mut wire2 = [0u8; 64];
    let n2 = narrow_in.serialize(&mut wire2).expect("serialize narrow");
    // Poison the destination so a missed zero-extension is visible rather than an incidental zero.
    let mut wide_out = WidHolder {
        inner: WidInner { x: 0xEE, y: 0xEE },
        tail: 0xEE,
    };
    wide_out
        .deserialize(&wire2[..n2])
        .expect("deserialize wide");
    assert_eq!(wide_out.inner.x, 5, "zero-ext inner.x");
    assert_eq!(wide_out.inner.y, 0, "zero-ext inner.y");
    assert_eq!(wide_out.tail, 77, "zero-ext tail");
    println!("zeroext_ok");
}

File: test/integration/ForwardCompatMain.go
//===----------------------------------------------------------------------===//
//
// Part of the OpenCyphal project, under the MIT licence
// SPDX-License-Identifier: MIT
//
//===----------------------------------------------------------------------===//

// Runtime forward/backward compatibility test for the generated Go serializers. See
// ForwardCompatDriver.c for the rationale: round-trip tests only decode a buffer with the same type
// version that wrote it, so the delimited version-skew skip is never exercised. This decodes across two
// versions of a delimited composite (wire.nar.Inner has one field; wire.wid.Inner appends a second).
package main

import (
	"fmt"
	"os"

	nar "dsdlfwdcompat/wire/nar"
	wid "dsdlfwdcompat/wire/wid"
)

func fail(msg string, code int) {
	fmt.Println(msg)
	os.Exit(code)
}

func main() {
	// Forward compatibility: wide writer -> narrow reader skips the appended Y.
	wideIn := wid.Holder@V1_0@{Inner: wid.Inner@V1_0@{X: 1, Y: 2}, Tail: 99}
	buf := make([]byte, 64)
	rc, n := wideIn.Serialize(buf)
	if rc != 0 {
		fail("go_serialize_wide_failed", 2)
	}
	var narrowOut nar.Holder@V1_0@
	rc2, _ := narrowOut.Deserialize(buf[:n])
	if rc2 != 0 {
		fail("go_deserialize_narrow_failed", 3)
	}
	if narrowOut.Inner.X != 1 || narrowOut.Tail != 99 {
		fail(fmt.Sprintf("go_forward_compat_mismatch x=%d tail=%d", narrowOut.Inner.X, narrowOut.Tail), 4)
	}
	fmt.Println("fwdcompat_ok")

	// Zero extension: narrow writer -> wide reader zero-fills the absent Y.
	narrowIn := nar.Holder@V1_0@{Inner: nar.Inner@V1_0@{X: 5}, Tail: 77}
	buf2 := make([]byte, 64)
	rc3, n3 := narrowIn.Serialize(buf2)
	if rc3 != 0 {
		fail("go_serialize_narrow_failed", 5)
	}
	// Poison the destination so a missed zero-extension is visible rather than an incidental zero.
	wideOut := wid.Holder@V1_0@{Inner: wid.Inner@V1_0@{X: 0xEE, Y: 0xEE}, Tail: 0xEE}
	rc4, _ := wideOut.Deserialize(buf2[:n3])
	if rc4 != 0 {
		fail("go_deserialize_wide_failed", 6)
	}
	if wideOut.Inner.X != 5 || wideOut.Inner.Y != 0 || wideOut.Tail != 77 {
		fail(fmt.Sprintf("go_zero_extension_mismatch x=%d y=%d tail=%d", wideOut.Inner.X, wideOut.Inner.Y, wideOut.Tail), 7)
	}
	fmt.Println("zeroext_ok")
}
